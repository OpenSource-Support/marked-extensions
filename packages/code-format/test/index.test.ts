/// <reference types="vitest/globals" />

import { Marked } from 'marked'
import prettierSvelte from 'prettier-plugin-svelte'
import markedCodeFormat from '../src/index.js'

it('should format code blocks using Prettier with default options', async () => {
  const md =
    '\n```js prettier\nfoo(reallyLongArg(), omgSoManyParameters(), IShouldRefactorThis(), isThereSeriouslyAnotherOne());\n```'

  const html = await new Marked().use(markedCodeFormat()).parse(md)

  expect(String(html)).toMatchSnapshot()
})

it('should format code blocks using Prettier with specified parser', async () => {
  const md =
    '\n```ts prettier="{ parser: \'typescript\' }"\n@observer class OrderLine {\n  @observable price: number = 0;\n}\n```'

  const html = await new Marked().use(markedCodeFormat()).parse(md)

  expect(String(html)).toMatchSnapshot()
})

it('should handle unsupported code block languages', async () => {
  const md =
    '\n```unsupported-language prettier\nconst user: { [key: string]: string \n| number | boolean } = {\n name: "John Doe", age: 30 };\n```'

  const html = await new Marked().use(markedCodeFormat()).parse(md)

  expect(String(html)).toMatchSnapshot()
})

it('should handle extended language mappings', async () => {
  const md = `
\`\`\`svelte prettier
<span><div>foo</div><span>bar</span></span>
<div pretend break>content</div>
\`\`\`
`

  const html = await new Marked()
    .use(
      markedCodeFormat({
        plugins: [prettierSvelte]
      })
    )
    .parse(md)

  expect(String(html)).toMatchSnapshot()
})

it('should handle formatting failures and preserve the original code block', async () => {
  // invalid TypeScript code to trigger a formatting failure
  const md = '```ts prettier\nconst x number = 1 // missing :\n```'

  const html = await new Marked().use(markedCodeFormat()).parse(md)

  expect(String(html)).toMatchSnapshot()
})

it('should not transform code blocks that do not have `prettier` attribute', async () => {
  const md =
    '\n```js\nfoo(reallyLongArg(), omgSoManyParameters(), IShouldRefactorThis(), isThereSeriouslyAnotherOne());\n```'

  const html = await new Marked().use(markedCodeFormat()).parse(md)

  expect(String(html)).toMatchSnapshot()
})

it('should be consumable by subsequent extensions', async () => {
  await new Marked()
    .use({ gfm: true })
    .use(markedCodeFormat())
    .use({
      walkTokens(token) {
        if (token.type !== 'code') return

        expect(token.lang).toMatch(/prettier/)
      }
    })
    .parse('```js prettier\n<Foo />\n```')
})
