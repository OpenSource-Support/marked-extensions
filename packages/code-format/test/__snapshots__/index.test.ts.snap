// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should format code blocks using Prettier with default options 1`] = `
"<pre><code class="language-js">foo(
  reallyLongArg(),
  omgSoManyParameters(),
  IShouldRefactorThis(),
  isThereSeriouslyAnotherOne()
)
</code></pre>
"
`;

exports[`should format code blocks using Prettier with specified parser 1`] = `
"<pre><code class="language-ts">@observer
class OrderLine {
  @observable price: number = 0
}
</code></pre>
"
`;

exports[`should handle extended language mappings 1`] = `
"<pre><code class="language-svelte">&lt;span
  &gt;&lt;div&gt;foo&lt;/div&gt;
  &lt;span&gt;bar&lt;/span&gt;&lt;/span
&gt;
&lt;div pretend break&gt;content&lt;/div&gt;
</code></pre>
"
`;

exports[`should handle formatting failures and preserve the original code block 1`] = `
"<pre><code class="language-ts">const x number = 1 // missing :
</code></pre>
"
`;

exports[`should handle unsupported code block languages 1`] = `
"<pre><code class="language-unsupported-language">const user: { [key: string]: string 
| number | boolean } = {
 name: &quot;John Doe&quot;, age: 30 };
</code></pre>
"
`;

exports[`should not transform code blocks that do not have \`prettier\` attribute 1`] = `
"<pre><code class="language-js">foo(reallyLongArg(), omgSoManyParameters(), IShouldRefactorThis(), isThereSeriouslyAnotherOne());
</code></pre>
"
`;
